### PR TITLE
[lldb] Convert LocateSymbolFile into a plugin

### PR DESCRIPTION
**Pull request by JDevlieghere**
This commit contains the initial scaffolding to convert the functionality currently implemented in LocateSymbolFile to a plugin architecture. The plugin approach allows us to easily add new ways to find plugins.

The current implementation abuses the host OS to use the DebugSymbols framework on macOS. With the plugin approach, we retain all the benefits (including the ability to compile it out) with a much cleaner level of abstraction.

To limit the scope of this patch, I've only converted LocateExecutableObjectFile. The goal is to do the same for the remaining LocateSymbolFile functions and eventually remove the file. To make reviewing easier, that will done as follow-up patches.

### CODE REVIEW

File: lldb/include/lldb/Symbol/SymbolLocator.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
+//===-- SymbolLocator.h -----------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLDB_SYMBOL_SYMBOLFILELOCATOR_H
```
Comment (bulbazord) [inline lldb/include/lldb/Symbol/SymbolLocator.h:9]: Include guard doesn't match file name

File: lldb/source/Plugins/SymbolLocator/DebugSymbols/SymbolLocatorDebugSymbols.cpp [hunk 2]
```diff
@@ -0,0 +1,328 @@
+//===-- SymbolLocatorDebugSymbols.cpp
+//----------------------------------------------===//
```
Comment (bulbazord) [inline lldb/source/Plugins/SymbolLocator/DebugSymbols/SymbolLocatorDebugSymbols.cpp:2]: Formatting

File: lldb/source/Plugins/SymbolLocator/DebugSymbols/SymbolLocatorDebugSymbols.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
+//===-- SymbolLocatorDebugSymbols.h ----------------------------------*- C++
+//-*-===//
```
Comment (bulbazord) [inline lldb/source/Plugins/SymbolLocator/DebugSymbols/SymbolLocatorDebugSymbols.h:2]: Formatting

File: lldb/source/Plugins/SymbolLocator/Default/SymbolLocatorDefault.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
+//===-- SymbolLocatorDefault.cpp
+//----------------------------------------------===//
```
Comment (bulbazord) [inline lldb/source/Plugins/SymbolLocator/Default/SymbolLocatorDefault.cpp:2]: Formatting

**Review by bulbazord — APPROVED** [review]
LGTM, all my comments were minor things I noticed. Thanks for doing this!

File: lldb/source/Plugins/SymbolLocator/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,4 @@
+add_subdirectory(Default)
+if (CMAKE_SYSTEM_NAME MATCHES "Darwin")
+  add_subdirectory(DebugSymbols)
```
Comment (kevinfrei) [inline lldb/source/Plugins/SymbolLocator/CMakeLists.txt:3]: I'm switching my previous PR to be a "Debuginfod" SymbolLocator plugin, and I think this plugin name is far too general. @clayborg tells me that  "DebugSymbols.framework" is a thing on MacOS, but maybe rename this "DebugSymbolsApple" or "MacOSDebugSymbols" or pretty much anything that would indicate to the random developer that this stuff is for (mac|i(Pad)|tv)OS/Apple/Darwin/Sherlock/dSYM symbol lookup.

**Pull request merged**